lib/logstash/codecs/cef.rb

```diff
@@ -181,6 +187,18 @@ def handle(data, &block)
     yield LogStash::Event.new("message" => data, "tags" => ["_cefparsefailure"])
   end
 
+  # mutation-safe version of `String#force_encoding` that reverses itself if
+  # the result does not produce a string with valid encoding.
+  def safe_coerce_encoding(data, target_encoding)
```

Comment:
@yaauie maybe we could add this functionality to [LogStash::Util::Charset](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/util/charset.rb)?

Also, maybe we can do now with the normal convert from Charset? It won't fail the decode as in this PR, but prints a warning:

```
irb(main):007:0> charset = LogStash::Util::Charset.new(Encoding::UTF_8)
=> #<LogStash::Util::Charset:0x11f5a842 @charset=#<Encoding:UTF-8>, @charset_encoding=#<Encoding:UTF-8>>
irb(main):021:0> str
=> "CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|src=192.168.1.11 target=aaaaa\x{82A0} \x{82A0}\x{82A0}aaaa msg=Description Omitted"
irb(main):022:0> ret = charset.convert(str)
[2018-05-30T15:13:47,098][WARN ][test                      ] Received an event that has a different character encoding than you configured. {:text=>"CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|src=192.168.1.11 target=aaaaa\\x82\\xA0 \\x82\\xA0\\x82\\xA0aaaa msg=Description Omitted", :expected_charset=>#<Encoding:UTF-8>}
=> "CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|src=192.168.1.11 target=aaaaa\\x82\\xA0 \\x82\\xA0\\x82\\xA0aaaa msg=Description Omitted"
irb(main):023:0> puts str
CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|src=192.168.1.11 target=aaaaa?? ????aaaa msg=Description Omitted
```